### PR TITLE
feat: add in-memory lock

### PR DESCRIPTION
**Pull request by stulzq**
**What this PR does**:
add lock in memory
**Which issue(s) this PR fixes**:

None
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: components/lock/in-memory/in_memory_lock.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+func (s *InMemoryLock) TryLock(req *lock.TryLockRequest) (*lock.TryLockResponse, error) {
+	item, ok := s.data.Load(req.ResourceId)
+	if !ok {
+		newItem := &lockData{
+			key:  req.ResourceId,
+			lock: &atomic.Int32{},
+		}
+		s.wLock.Lock()
```
Comment (seeflood) [inline components/lock/in-memory/in_memory_lock.go:61]: I remember that sync.Map is concurrency-safe. There seems no need to use a lock?

**Review by seeflood — CHANGES_REQUESTED** [review]

File: components/lock/in-memory/in_memory_lock.go (hunk 1 above)
Comment (stulzq) [reply components/lock/in-memory/in_memory_lock.go:61]: > I remember that sync.Map is concurrency-safe. There seems no need to use a lock?

Line 73-78, has delete and store operation, so I locked it here.

File: components/lock/in-memory/in_memory_lock.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			Status: lock.SUCCESS,
+		}, nil
+	}
+
+	if d.owner != req.LockOwner {
+		return &lock.UnlockResponse{
+			Status: lock.LOCK_BELONG_TO_OTHERS,
+		}, nil
+	}
+
```
Comment (seeflood) [inline components/lock/in-memory/in_memory_lock.go:126]: Oh,it's not success here

**Review by seeflood — CHANGES_REQUESTED** [review]

File: components/lock/in-memory/in_memory_lock.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	if !d.lock.CAS(1, 0) {
+		return &lock.UnlockResponse{
+			Status: lock.LOCK_UNEXIST,
+		}, nil
+	}
+
+	d.owner = ""
```
Comment (seeflood) [inline components/lock/in-memory/in_memory_lock.go:133]: 我用中文吧，这可能有并发问题
比如：
协程1 执行了127行成功
协程2执行96行成功，成功上锁、正在写owner信息
协程1执行133行，把协程2写的覆盖掉。

用cas比较tricky、容易出错，除了上述情况还可能有ABA问题，所以简单点可以改成用锁

**Review by seeflood — CHANGES_REQUESTED** [review]

File: components/lock/in-memory/in_memory_lock.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+	//0 unlock, 1 lock
+	d := item.(*lockData)
+
+	// 2. Construct a new one if the lockData has expired
+	//check expire
+	if d.owner != "" && time.Now().Before(d.expireTime) {
+		s.wLock.Lock()
+		//double check
+		s.data.Delete(req.ResourceId)
```
Comment (seeflood) [inline components/lock/in-memory/in_memory_lock.go:78]: I suppose we need to double check before delete

File: components/lock/in-memory/in_memory_lock.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		return &lock.UnlockResponse{
+			Status: lock.LOCK_UNEXIST,
+		}, nil
+	}
+
+	d := item.(*lockData)
+	// 2. check the owner information
+	if d.lock.Load() != 1 {
+		return &lock.UnlockResponse{
+			Status: lock.SUCCESS,
```
Comment (seeflood) [inline components/lock/in-memory/in_memory_lock.go:117]: Oh,it's not success here

**Review by seeflood — CHANGES_REQUESTED** [review]

**Review by seeflood — APPROVED** [review]
lgtm. I added some comments

**Review by zhenjunMa — DISMISSED** [review]

**Review by seeflood — DISMISSED** [review]
lgtm. I just fixed a bug

**Review by seeflood — APPROVED** [review]
lgtm

**Review by zhenjunMa — APPROVED** [review]

**Comment by seeflood** [comment]
This component can make debug easier.Thanks!

**Pull request merged**